### PR TITLE
Handle advanced scenarios for package updates

### PR DESCRIPTION
**Pull request by mortent**
This addresses an issue where new subpackages added in upstream where not added correctly. It also incorporates the cleanup of how we update the Kptfile from https://github.com/GoogleContainerTools/kpt/pull/1630, but doesn't make any changes to the pipeline.

It also continues the effort to consolidate code in `internal/pkg` and `internal/util/pkgutil`. This adds a new function for finding subpackages in the `internal/pkg` package that replaces all the variants we had for this.

### CODE REVIEW

File: internal/pkg/pkg.go [hunk 1]
```diff
@@ -145,14 +144,63 @@ func (p *Pkg) RelativePathTo(ancestorPkg *Pkg) (string, error) {
...
+)
+
+// Subpackages returns a slice of paths to any subpackages of the provided path.
+// The matcher parameter decides if all types of subpackages should be considered,
+// and the recursive parameter determines if only direct subpackages are
+// considered. All returned paths will be relative to the provided rootPath.
+// The top level package is not considered a subpackage. If the provided path
+// doesn't exist, an empty slice will be returned.
+// Symlinks are ignored.
+func Subpackages(rootPath string, matcher SubpackageMatcher, recursive bool) ([]string, error) {
```
Comment (frankfarzan) [inline internal/pkg/pkg.go:188]: Do you want to export this function? It uses raw string for input and output instead of specific types,.

Comment (mortent) [reply internal/pkg/pkg.go:188]: Yeah, at least for now I need it to just accept strings. The update code needs to use this functionality on clones of git repos in temporary directories, so the `DisplayPath` functionality of package doesn't work all that well. It also involves traversing multiple directory hierarchies in parallel, so it needs to manipulate paths. 
So this isn't perfect, but I think it is a step in the right direction. This gives up one tested implementation of looking up subpackages. Will continue to see if we can rely more on the packaging abstraction.

Comment (frankfarzan) [reply internal/pkg/pkg.go:188]: Add a TODO for ideal end-state?

Comment (mortent) [reply internal/pkg/pkg.go:190]: Added a TODO.

File: internal/pkg/pkg_test.go [hunk 2]
```diff
@@ -307,3 +310,188 @@ func TestSubpackages(t *testing.T) {
...
+		matcher   []SubpackageMatcher
+		recursive []bool
+		expected  []string
+	}
+
+	testCases := map[string]struct {
+		pkg   *pkgbuilder.RootPkg
+		cases []variants
+	}{
+		"remote and local nested subpackages": {
```
Comment (phanimarupaka) [inline internal/pkg/pkg_test.go:326]: Can you comment the `tree` output of the package for readability. Nested structures in test cases are tough to read.

File: internal/pkg/pkg.go [hunk 3]
```diff
@@ -145,14 +144,65 @@ func (p *Pkg) RelativePathTo(ancestorPkg *Pkg) (string, error) {
 // TODO: This does not support symlinks, so we need to figure out how
 // we should support that with kpt.
 func (p *Pkg) DirectSubpackages() ([]*Pkg, error) {
```
Comment (phanimarupaka) [inline internal/pkg/pkg.go:146]: A subpackage is a direct subpackage iff it exists in the top level folder of a package ? What if it exists within the directory ? We will be missing out such packages when do hydration using functions.

This equation must hold true.
Recursive sum of all direct subpackages = Total number of packages

File: internal/util/update/update.go [hunk 4]
```diff
@@ -124,21 +135,21 @@ func (u Command) Run() error {
...
 	}
 	if strings.TrimSpace(g.Stdout.String()) != "" {
-		return errors.Errorf("must commit package %s to git before attempting to update",
+		return fmt.Errorf("must commit package %s to git before attempting to update",
 			u.Pkg.UniquePath.String())
 	}
 
 	rootKf, err := u.Pkg.Kptfile()
 	if err != nil {
-		return errors.Errorf("unable to read package Kptfile: %w", err)
```
Comment (phanimarupaka) [inline internal/util/update/update.go:137]: Why this change ? Aren't we using `errors.Errorf` elsewhere?

File: internal/pkg/pkg_test.go (hunk 2 above)
Comment (mortent) [reply internal/pkg/pkg_test.go:326]: Added an illustration for each test case that shows the package hierarchy. I think this should make it easier to quickly see the structure.

File: internal/pkg/pkg.go (hunk 3 above)
Comment (mortent) [reply internal/pkg/pkg.go:148]: Yeah, this does find subpackages nested in directories. Direct here means it doesn't include subpackages nested inside other subpackages (so sub-subpackages). I have updated the description to make this clear.

File: internal/util/update/update.go (hunk 4 above)
Comment (mortent) [reply internal/util/update/update.go:137]: We are in the process of moving to a slightly different way to create errors: https://github.com/GoogleContainerTools/kpt/pull/1639. But those should be addressed in separate PRs, so I have reverted the change here to keep using the `errors.Errorf` package.

File: internal/pkg/pkg_test.go [hunk 5]
```diff
@@ -307,3 +310,204 @@ func TestSubpackages(t *testing.T) {
...
+			//  │       └── remote-sub3 (remote)
+			//  │           └── Kptfile
+			//  └── local-sub1 (local)
+			//      ├── Kptfile
+			//      ├── directory
+			//      │   └── remote-sub3 (remote)
+			//      │       └── Kptfile
+			//      └── local-sub2 (local)
+			//          └── Kptfile
+			pkg: pkgbuilder.NewRootPkg().
```
Comment (phanimarupaka) [inline internal/pkg/pkg_test.go:340]: Thanks for adding the structure. It is clear now. 

On a separate note, there was a discussion on other functions PR to create tests with data so that anyone can add more tests without needing to learn about pkgbuilder. https://github.com/GoogleContainerTools/kpt/tree/next/e2e/testdata/fn-render/basicpipeline
We should do that eventually in future.

**Review by phanimarupaka — APPROVED** [review]
